python/lsst/sims/maf/metrics/seasonMetrics.py

```diff
@@ -0,0 +1,245 @@
+        units = 'days'
+        self.mjdCol = mjdCol
+        self.expTimeCol = expTimeCol
+        self.minExpTime = minExpTime
+        self.reduceFunc = reduceFunc
+        super().__init__(col=[self.mjdCol, self.expTimeCol],
+                         units=units, metricName=metricName, **kwargs)
+
+    def run(self, dataSlice, slicePoint):
+        """Calculate the (reduceFunc) of the length of each season.
+        Uses the slicePoint RA/Dec to calculate the position in question, then uses the times of the visits
+        to assign them into seasons (based on where the sun is relative to the slicePoint RA).
+
+        Parameters
+        ----------
+        dataSlice : numpy.array
+            Numpy structured array containing the data related to the visits provided by the slicer.
+        slicePoint : dict
+            Dictionary containing information about the slicepoint currently active in the slicer.
+
+        Returns
+        -------
+        float
+           The (reduceFunc) of the length of each season, in days.
+        """
+        # Order data Slice/times and exclude visits which are too short.
+        long = np.where(dataSlice[self.expTimeCol] > self.minExpTime)
+        data = np.sort(dataSlice[long], order=self.mjdCol)
+        # SlicePoints ra/dec are always in radians - convert to degrees to calculate season
+        seasons = calcSeason(np.degrees(slicePoint['ra']), data[self.mjdCol])
+        firstOfSeason, lastOfSeason = findSeasonEdges(seasons)
+        seasonlengths = data[self.mjdCol][lastOfSeason] - data[self.mjdCol][firstOfSeason]
+        result = self.reduceFunc(seasonlengths)
+        return result
+
+
+class CampaignLengthMetric(BaseMetric):
+    """Calculate the number of seasons (roughly, years) a pointing is observed for.
+    This corresponds to the 'campaign length' for lensed quasar time delays.
+    """
+    def __init__(self, mjdCol='observationStartMJD', expTimeCol='visitExposureTime', minExpTime=20, **kwargs):
+        units = ''
+        self.expTimeCol = expTimeCol
+        self.minExpTime = minExpTime
+        self.mjdCol = mjdCol
+        super().__init__(col=[self.mjdCol, self.expTimeCol], units=units, **kwargs)
+
+    def run(self, dataSlice, slicePoint):
+        # Order data Slice/times and exclude visits which are too short.
+        long = np.where(dataSlice[self.expTimeCol] > self.minExpTime)
+        data = np.sort(dataSlice[long], order=self.mjdCol)
+        seasons = calcSeason(np.degrees(slicePoint['ra'], data[self.mjdCol]))
+        intSeasons = np.floor(seasons)
+        count = len(np.unique(intSeasons))
+        return count
+
+
+class MeanCampaignFrequencyMetric(BaseMetric):
+    """Calculate the mean separation between nights, within a season - then the mean over the campaign.
+    Calculate per season, to avoid any influence from season gaps.
+    """
+    def __init__(self, mjdCol='observationStartMJD', expTimeCol='visitExposureTime', minExpTime=20,
```

Comment:
A `minExpTime` is only a proxy and can still fail in twilight or bright time. Probably better to set a fiducial magnitude and SNR limit. I usually default to r=22 and SNR=5, then use the fivesigmadepth to cut the shallow observations.

Reply:
Yeah, I get that. This was kind of the way they wanted to go, rather than setting a strict magnitude limit. I will check with Timo if r=22 seems reasonable for their objects. They're using visits in all filters, so I'd also need their colors.